packages/core/src/2d/text/TextUtils.ts

```diff
@@ -0,0 +1,413 @@
+
+        if (top === -1) {
+          top = y;
+        }
+
+        if (left === -1 || x < left) {
+          left = x;
+        }
+
+        if (right === null || x > right) {
+          right = x;
+        }
+
+        if (bottom === null || y > bottom) {
+          bottom = y;
+        }
+      }
+    }
+
+    if (top !== -1) {
+      top = Math.max(0, top - 1);
+      bottom = Math.min(height - 1, bottom + 1);
+      left = Math.max(0, left - 1);
+      right = Math.min(width - 1, right + 1);
+      width = right - left + 1;
+      height = bottom - top + 1;
+      data = context.getImageData(left, top, width, height);
+    }
+
+    return {
+      width,
+      height,
+      data
+    };
+  }
+
+  /**
+   * Update text.
+   * @param textMetrics - the text metrics object
+   * @param fontStr - the font string
+   * @param horizontalAlignment - the horizontal alignment
+   * @param verticalAlignment - the vertical alignment
+   */
+  public static updateText(
+    textMetrics: TextMetrics,
+    fontStr: string,
+    horizontalAlignment: TextHorizontalAlignment,
+    verticalAlignment: TextVerticalAlignment
+  ): void {
+    const { canvas, context } = TextUtils.textContext();
+    const { width, height } = textMetrics;
+    // reset canvas's width and height.
+    canvas.width = width;
+    canvas.height = height;
+    // clear canvas.
+    context.font = fontStr;
+    context.clearRect(0, 0, width, height);
+    // set canvas font info.
+    context.textBaseline = "top";
+    context.fillStyle = "#fff";
+
+    // draw lines.
+    const { lines, lineHeight, lineWidths } = textMetrics;
+    for (let i = 0, l = lines.length; i < l; ++i) {
+      const lineWidth = lineWidths[i];
+      const pos = TextUtils._tempVec2;
+      TextUtils._calculateLinePosition(
+        width,
+        height,
+        lineWidth,
+        lineHeight,
+        i,
+        l,
+        horizontalAlignment,
+        verticalAlignment,
+        pos
+      );
+      const { x, y } = pos;
+      if (y + lineHeight >= 0 && y < height) {
+        // +1, for chrome.
+        context.fillText(lines[i], x, y + 1);
+      }
+    }
+  }
+
+  /**
+   * Update canvas with the data.
+   * @param width - the new width of canvas
+   * @param height - the new height of canvas
+   * @param data - the new data of canvas
+   * @returns the canvas after update
+   */
+  public static updateCanvas(width: number, height: number, data: ImageData): HTMLCanvasElement | OffscreenCanvas {
+    const { canvas, context } = TextUtils.textContext();
+    canvas.width = width;
+    canvas.height = height;
+    context.putImageData(data, 0, 0);
+    return canvas;
+  }
+
+  private static _wordWrap(text: string, width: number, enableWrapping: boolean, fontStr: string): Array<string> {
+    const { context } = TextUtils.textContext();
+    const { _maxWidth: maxWidth } = TextUtils;
+    const widthInPixel = width * TextUtils._pixelsPerUnit;
+    const output = new Array<string>();
```

Comment:
`DisorderedArray` can be used to save memory.

Reply:
Only add element, not need to use DisorderedArray